src/Agent.Worker/JobRunner.cs

```diff
@@ -10,100 +10,149 @@ namespace Microsoft.VisualStudio.Services.Agent.Worker
                     .Where(x => string.Equals(x.HostType, hostType, StringComparison.OrdinalIgnoreCase))
                     .ToArray();
 
                 // Add the prepare steps.
+                Trace.Info("Adding job prepare extensions.");
                 List<IStep> steps = new List<IStep>();
                 foreach (IJobExtension extension in extensions)
                 {
                     if (extension.PrepareStep != null)
                     {
                         Trace.Verbose($"Adding {extension.GetType().Name}.{nameof(extension.PrepareStep)}.");
-                        extension.PrepareStep.ExecutionContext = jobExecutionContext.CreateChild(Guid.NewGuid(), extension.PrepareStep.DisplayName);
+                        extension.PrepareStep.ExecutionContext = jobContext.CreateChild(Guid.NewGuid(), extension.PrepareStep.DisplayName);
                         steps.Add(extension.PrepareStep);
                     }
                 }
 
                 // Add the task steps.
+                Trace.Info("Adding tasks.");
                 foreach (TaskInstance taskInstance in message.Tasks)
                 {
                     Trace.Verbose($"Adding {taskInstance.DisplayName}.");
                     var taskRunner = HostContext.CreateService<ITaskRunner>();
-                    taskRunner.ExecutionContext = jobExecutionContext.CreateChild(taskInstance.InstanceId, taskInstance.DisplayName);
+                    taskRunner.ExecutionContext = jobContext.CreateChild(taskInstance.InstanceId, taskInstance.DisplayName);
                     taskRunner.TaskInstance = taskInstance;
                     steps.Add(taskRunner);
                 }
 
                 // Add the finally steps.
+                Trace.Info("Adding job finally extensions.");
                 foreach (IJobExtension extension in extensions)
                 {
                     if (extension.FinallyStep != null)
                     {
                         Trace.Verbose($"Adding {extension.GetType().Name}.{nameof(extension.FinallyStep)}.");
-                        extension.FinallyStep.ExecutionContext = jobExecutionContext.CreateChild(Guid.NewGuid(), extension.FinallyStep.DisplayName);
+                        extension.FinallyStep.ExecutionContext = jobContext.CreateChild(Guid.NewGuid(), extension.FinallyStep.DisplayName);
                         steps.Add(extension.FinallyStep);
                     }
                 }
 
                 // Download tasks if not already in the cache
-                var taskManager = HostContext.GetService<ITaskManager>();
-                await taskManager.EnsureTasksExist(message.Tasks);
+                Trace.Info("Downloading task definitions.");
+                var taskDefinitionManager = HostContext.GetService<ITaskDefinitionManager>();
+                try
+                {
+                    await taskDefinitionManager.DownloadAsync(jobContext, message.Tasks);
```

Comment:
1) I renamed TaskManager to TaskDefinitionManager.
2) I renamed EnsureTasksExist to DownloadAsync.
3) I added a catch block here so we understand at a high level where the error came from.